apps/hash-ai-worker-ts/src/activities/flow-activities/research-entities-action/link-follower-agent/get-link-follower-next-tool-calls.ts

```diff
@@ -0,0 +1,317 @@
+import type { LocalEntitySummary } from "../../shared/infer-facts-from-text/get-entity-summaries-from-text";
+import type { Fact } from "../../shared/infer-facts-from-text/types";
+import type { Link } from "./extract-links-from-content";
+
+const defaultModel: LlmParams["model"] = "claude-3-5-sonnet-20240620";
+
+const getLinkFollowerNextToolCallsSystemPrompt = dedent(`
+  You are a link follower agent.
+
+  <UserMessage>
+  The user will provide you with:
+    - Task: a research task you have been instructed to fulfill,
+        based on the contents of a resource (e.g. a webpage)
+    - Previously Visited Links: links to resources which have been
+        previously visited to extract facts
+    - Entities: a list of entities for which facts have been gathered, including:
+        - name: the name of the entity
+        - summary: a summary of the entity
+        - entityType: the type of the entity
+        - facts: the facts that have been gathered about the entity
+    - Possible Next Links: a list of the possible next links which
+        may be explored next to gather more facts and fulfill the task
+  </UserMessage>
+
+  <TaskDescription>
+  Using the provided tools, you must make a decision on what to do next to fulfill the task.
+
+  You can only make a single tool call, which must be one of the following:
+    - exploreLinks: call this tool to explore additional links to gather more facts that may fulfill the task
+    - complete: complete the research task if all the gathered facts fulfill the task
+    - terminate: terminate the research task if it cannot be progressed further
+  </TaskDescription>
+`);
+
+type GetLinkFollowerNextToolCallsParams = {
+  task: string;
+  entitySummaries: LocalEntitySummary[];
+  factsGathered: Fact[];
+  previouslyVisitedLinks: { url: string }[];
+  possibleNextLinks: Link[];
+};
+
+const simplifyFactForLlmConsumption = (fact: Fact) => {
+  return `${fact.text} ${fact.prepositionalPhrases.join(", ")}`;
+};
+
+const generateUserMessage = (
+  params: GetLinkFollowerNextToolCallsParams,
+): LlmUserMessage => {
+  const {
+    task,
+    entitySummaries,
+    factsGathered,
+    previouslyVisitedLinks,
+    possibleNextLinks,
+  } = params;
+
+  return {
+    role: "user",
+    content: [
+      {
+        type: "text",
+        text: dedent(`
+Task: ${task}
+Previously Visited Links: ${JSON.stringify(previouslyVisitedLinks)}
+Entities: ${JSON.stringify(
+          entitySummaries.map(({ localId, name, summary, entityTypeId }) => {
+            const factsAboutEntity = factsGathered.filter(
+              (fact) => fact.subjectEntityLocalId === localId,
+            );
+
+            return {
+              name,
+              summary,
+              entityType: entityTypeId,
+              facts: factsAboutEntity,
+            };
+          }),
+        )}
+Facts Gathered: ${JSON.stringify(factsGathered.map(simplifyFactForLlmConsumption))}
```

Comment:
Isn't `Facts Gathered` redundant, as the facts are also included in the `Entities`?

Reply:
Thanks for catching, forgot to remove this when I moved it into the entities list. Have removed it in https://github.com/hashintel/hash/pull/4649/commits/e83ee1803450d9e02afdfc63939652ab2f4094e0